### PR TITLE
Add metadata to extracted audio file too.

### PR DESCRIPTION
**Pull request by NextLight**
This fixes metadata not being present if not supported by the initial container (i.e. artist attribute in webm).

### CODE REVIEW

**Comment by dstftw** [comment]
This causes metadata embedding to be completely skipped in default scenario.
With this change (no metadata is embedded):
```
[debug] System config: []
[debug] User config: [u'-o', u'%(title)s-%(id)s-f%(format_id)s.%(ext)s']
[debug] Command-line args: [u'BaW_jenozKc', u'--add-metadata', u'-v']
[debug] Encodings: locale cp1251, fs mbcs, out cp866, pref cp1251
[debug] youtube-dl version 2016.06.03
[debug] Git HEAD: 302d839
[debug] Python version 2.6.2 - Windows-post2008Server-6.2.9200
[debug] exe versions: ffmpeg 3.0, ffprobe 3.0, rtmpdump 2.4
[debug] Proxy map: {}
[youtube] BaW_jenozKc: Downloading webpage
[youtube] BaW_jenozKc: Downloading video info webpage
[youtube] BaW_jenozKc: Extracting video information
[youtube] BaW_jenozKc: Downloading MPD manifest
WARNING: Requested formats are incompatible for merge and will be merged into mkv.
[debug] Invoking downloader on u'https://r3---sn-ug5onuxaxjvh-v8cl.googlevideo.com/videoplayback?id=05a5bf8de9e8cca7&itag=137&source=youtube&requiressl=yes&initcwndbps=3520000&mm=31&mn=sn-ug5onuxaxjvh-v8cl&ms=au&pl=20&mv=m&ratebypass=yes&mime=video/mp4&gir=yes&clen=2208750&lmt=1387961822987808&dur=9.800&key=dg_yt0&mt=1465417645&upn=BTL2KorbpAA&sver=3&signature=5B476172B1643638689C1F36F9D6A225E0BC6709.448F3E992D55889E25895B1740232F73346DDA04&fexp=9406006,9408208,9416126,9416891,9422596,9428398,9428741,9431012,9431890,9432058,9433043,9433096,9433380,9433946,9434604,9435527,9435876,9435991,9436224,9436618,9437066,9437257,9437553,9437656,9437917,9438280,9438568&ip=109.171.17.81&ipbits=0&expire=1465439528&sparams=ip,ipbits,expire,id,itag,source,requiressl,initcwndbps,mm,mn,ms,pl,mv,ratebypass,mime,gir,clen,lmt,dur'
[download] Destination: youtube-dl test video ''_-BaW_jenozKc-f137.f137.mp4
[download] 100% of 2.11MiB in 00:00
[debug] Invoking downloader on u'https://r3---sn-ug5onuxaxjvh-v8cl.googlevideo.com/videoplayback?id=05a5bf8de9e8cca7&itag=251&source=youtube&requiressl=yes&initcwndbps=3520000&mm=31&mn=sn-ug5onuxaxjvh-v8cl&ms=au&pl=20&mv=m&ratebypass=yes&mime=audio/webm&gir=yes&clen=190742&lmt=1422348740429751&dur=9.841&key=dg_yt0&mt=1465417645&upn=BTL2KorbpAA&sver=3&signature=2F5E1F8E33A8A355FC4D3EEE1F8F1549CF087957.73786DA6E5515848E0F1E6664C077918A695D6D8&fexp=9406006,9408208,9416126,9416891,9422596,9428398,9428741,9431012,9431890,9432058,9433043,9433096,9433380,9433946,9434604,9435527,9435876,9435991,9436224,9436618,9437066,9437257,9437553,9437656,9437917,9438280,9438568&ip=109.171.17.81&ipbits=0&expire=1465439528&sparams=ip,ipbits,expire,id,itag,source,requiressl,initcwndbps,mm,mn,ms,pl,mv,ratebypass,mime,gir,clen,lmt,dur'
[download] Destination: youtube-dl test video ''_-BaW_jenozKc-f251.f251.webm
[download] 100% of 186.27KiB in 00:00
[ffmpeg] Merging formats into "youtube-dl test video ''_-BaW_jenozKc-f137+251.mkv"
[debug] ffmpeg command line: ffmpeg -y -i "file:youtube-dl test video ''_-BaW_jenozKc-f137.f137.mp4" -i "file:youtube-dl test video ''_-BaW_jenozKc-f251.f251.webm" -c copy -map 0:v:0 -map 1:a:0 "file:youtube-dl test video ''_-BaW_jenozKc-f137+251.temp.mkv"
Deleting original file youtube-dl test video ''_-BaW_jenozKc-f137.f137.mp4 (pass -k to keep)
Deleting original file youtube-dl test video ''_-BaW_jenozKc-f251.f251.webm (pass -k to keep)
```
Without (correct):
```
[debug] System config: []
[debug] User config: [u'-o', u'%(title)s-%(id)s-f%(format_id)s.%(ext)s']
[debug] Command-line args: [u'BaW_jenozKc', u'--add-metadata', u'-v']
[debug] Encodings: locale cp1251, fs mbcs, out cp866, pref cp1251
[debug] youtube-dl version 2016.06.03
[debug] Git HEAD: 411c590
[debug] Python version 2.6.2 - Windows-post2008Server-6.2.9200
[debug] exe versions: ffmpeg 3.0, ffprobe 3.0, rtmpdump 2.4
[debug] Proxy map: {}
[youtube] BaW_jenozKc: Downloading webpage
[youtube] BaW_jenozKc: Downloading video info webpage
[youtube] BaW_jenozKc: Extracting video information
[youtube] BaW_jenozKc: Downloading MPD manifest
WARNING: Requested formats are incompatible for merge and will be merged into mkv.
[debug] Invoking downloader on u'https://r3---sn-ug5onuxaxjvh-v8cl.googlevideo.com/videoplayback?id=05a5bf8de9e8cca7&itag=137&source=youtube&requiressl=yes&mm=31&mn=sn-ug5onuxaxjvh-v8cl&mv=m&ms=au&initcwndbps=3520000&pl=20&ratebypass=yes&mime=video/mp4&gir=yes&clen=2208750&lmt=1387961822987808&dur=9.800&mt=1465417645&key=dg_yt0&sver=3&signature=33EC9D26135D743B6C14B4E8F94A38B074F5ADC2.520E57543AF527ABDB950BA0C1F166A2452E641C&upn=ZQpSkRdA254&fexp=9416126,9416891,9418777,9422596,9428398,9431012,9432684,9433096,9433380,9433946,9434000,9435527,9435693,9435876,9436049,9436221,9436883,9437066,9437177,9437191,9437553,9437597,9438327,9438362,9438568,9438663&ip=109.171.17.81&ipbits=0&expire=1465439570&sparams=ip,ipbits,expire,id,itag,source,requiressl,mm,mn,mv,ms,initcwndbps,pl,ratebypass,mime,gir,clen,lmt,dur'
[download] Destination: youtube-dl test video ''_-BaW_jenozKc-f137.f137.mp4
[download] 100% of 2.11MiB in 00:00
[debug] Invoking downloader on u'https://r3---sn-ug5onuxaxjvh-v8cl.googlevideo.com/videoplayback?id=05a5bf8de9e8cca7&itag=251&source=youtube&requiressl=yes&mm=31&mn=sn-ug5onuxaxjvh-v8cl&mv=m&ms=au&initcwndbps=3520000&pl=20&ratebypass=yes&mime=audio/webm&gir=yes&clen=190742&lmt=1422348740429751&dur=9.841&mt=1465417645&key=dg_yt0&sver=3&signature=38EDD70EB68862E856550B78F50903B3D5EB49D3.7E4F7B1EDBFA68C6D61238C2BF5BCB5682DE5097&upn=ZQpSkRdA254&fexp=9416126,9416891,9418777,9422596,9428398,9431012,9432684,9433096,9433380,9433946,9434000,9435527,9435693,9435876,9436049,9436221,9436883,9437066,9437177,9437191,9437553,9437597,9438327,9438362,9438568,9438663&ip=109.171.17.81&ipbits=0&expire=1465439570&sparams=ip,ipbits,expire,id,itag,source,requiressl,mm,mn,mv,ms,initcwndbps,pl,ratebypass,mime,gir,clen,lmt,dur'
[download] Destination: youtube-dl test video ''_-BaW_jenozKc-f251.f251.webm
[download] 100% of 186.27KiB in 00:00
[ffmpeg] Merging formats into "youtube-dl test video ''_-BaW_jenozKc-f137+251.mkv"
[debug] ffmpeg command line: ffmpeg -y -i "file:youtube-dl test video ''_-BaW_jenozKc-f137.f137.mp4" -i "file:youtube-dl test video ''_-BaW_jenozKc-f251.f251.webm" -c copy -map 0:v:0 -map 1:a:0 "file:youtube-dl test video ''_-BaW_jenozKc-f137+251.temp.mkv"
Deleting original file youtube-dl test video ''_-BaW_jenozKc-f137.f137.mp4 (pass -k to keep)
Deleting original file youtube-dl test video ''_-BaW_jenozKc-f251.f251.webm (pass -k to keep)
[ffmpeg] Adding metadata to 'youtube-dl test video ''_-BaW_jenozKc-f137+251.mkv'
[debug] ffmpeg command line: ffmpeg -y -i "file:youtube-dl test video ''_-BaW_jenozKc-f137+251.mkv" -c copy -metadata "comment=test chars:  \"'/\\
test URL: https://github.com/rg3/youtube-dl/iss...

This is a test video for youtube-dl.

For more information, contact phihag@phihag.de ." -metadata "description=test chars:  \"'/\\
test URL: https://github.com/rg3/youtube-dl/iss...

This is a test video for youtube-dl.

For more information, contact phihag@phihag.de ." -metadata 'artist=Philipp Hagemeister' -metadata "title=youtube-dl test video \"'/\\" -metadata date=20121002 -metadata 'purl=https://www.youtube.com/watch?v=BaW_jenozKc' "file:youtube-dl test video ''_-BaW_jenozKc-f137+251.temp.mkv"
```

**Comment by NextLight** [comment]
Whoops sorry, I'm retard. I fixed it. Feel free to merge them in a single commit if you want.

File: youtube_dl/__init__.py [hunk 1]
```diff
@@ -221,15 +221,17 @@ def parse_retries(retries):
...
+        if opts.addmetadata and opts.keepvideo:
+            postprocessors.append({'key': 'FFmpegMetadata'})
         postprocessors.append({
             'key': 'FFmpegExtractAudio',
             'preferredcodec': opts.audioformat,
             'preferredquality': opts.audioquality,
             'nopostoverwrites': opts.nopostoverwrites,
         })
+    if opts.addmetadata:
+        postprocessors.append({'key': 'FFmpegMetadata'})
```
Comment (dstftw) [inline youtube_dl/__init__.py]: This should probably go after `recodevideo` as well.

**Comment by dstftw** [comment]
Related #5594.

**Comment by NextLight** [comment]
Yep you are right, fix'd. Worst PR ever lol, I'm sorry.

File: youtube_dl/__init__.py [hunk 2]
```diff
@@ -221,9 +221,9 @@ def parse_retries(retries):
             'key': 'MetadataFromTitle',
             'titleformat': opts.metafromtitle
         })
-    if opts.addmetadata:
-        postprocessors.append({'key': 'FFmpegMetadata'})
     if opts.extractaudio:
+        if opts.addmetadata and opts.keepvideo:
+            postprocessors.append({'key': 'FFmpegMetadata'})
```
Comment (dstftw) [inline youtube_dl/__init__.py]: I guess similar code should go under `recodevideo` also.

Comment (NextLight) [inline youtube_dl/__init__.py]: And I guess also when merging DASHes.

**Comment by NextLight** [comment]
I figured this would be a nicest way. Hopefully it isn't too invasive.
Also this takes for granted that the file to be deleted is a 'media' (audio/video) file (if not ffmpeg will throw an error I guess), not sure if it's ok.

**Comment by Spenhouet** [comment]
This fix is still needed (see https://github.com/rg3/youtube-dl/issues/11116).

**Comment by imsys** [comment]
@NextLight's patch works nice, but as he said, he invokes the addmetadata action in the part of the script where it is deleting temp files.

I thought maybe instead of adding the metadata before and after converting the audio, maybe we could just add it after the conversion by just changing the postprocessors order in the `youtube_dl/__init__.py` file.


```python

    # PostProcessors
    postprocessors = []
    
    if opts.extractaudio:
        postprocessors.append({
            'key': 'FFmpegExtractAudio',
            'preferredcodec': opts.audioformat,
            'preferredquality': opts.audioquality,
            'nopostoverwrites': opts.nopostoverwrites,
        })
    if opts.metafromtitle:
        postprocessors.append({
            'key': 'MetadataFromTitle',
            'titleformat': opts.metafromtitle
        })
    if opts.addmetadata:
        postprocessors.append({'key': 'FFmpegMetadata'})
    if opts.recodevideo:
        postprocessors.append({
            'key': 'FFmpegVideoConvertor',
            'preferedformat': opts.recodevideo,
        })
```

**Comment by dstftw** [comment]
Since 54a3a8827baf71c553723b6766e676369e9c743c metadata is added after conversion.

**Pull request closed without merging**